server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ClusterBalanceStats.java

```diff
@@ -241,14 +242,19 @@ private static NodeBalanceStats createFrom(
                 }
             }
 
+            Double nodeWeight = desiredBalance.weightsPerNode().isEmpty()
+                ? null
+                : desiredBalance.weightsPerNode().get(routingNode.node()).nodeWeight();
```

Comment:
Two suggestions
1. We can add `assert desiredBalance != null` before this line to make the intention clear that it should not be `null`. Otherwise if it throws a NPE in some future tests, it would make people pause and wonder whether null handling is needed here.
2. Desired balance is compueted asynchrounously. It is possible that a new node from the latest cluster state is not yet included. So we need to count for it as well, e.g. something like:
```java
Double nodeWeight = Optional.of(desiredBalance.weightsPerNode().get(routingNode.node())
    .map(NodeWeightStats::nodeWeight)
    .orElse(null);
```